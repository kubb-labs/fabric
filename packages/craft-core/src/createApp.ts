import { defineApp } from './defineApp.ts'
import { ref } from './reactive/ref.ts'
import type * as KubbFile from './types'

export const createApp = defineApp<{ files: Array<KubbFile.File> }>((app) => {
  const files = ref<Array<KubbFile.File>>(app.files)
  const output = ref<string>('')

  return {
    files,
    output,
    async run() {
      throw new Error('Method not implemented')
    },
    async waitUntilExit() {
      throw new Error('Method not implemented')
    },
  }
})
