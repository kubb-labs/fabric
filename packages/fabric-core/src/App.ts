import type * as KubbFile from './KubbFile.ts'
import type { Plugin } from './plugins/types.ts'
import type { Parser } from './parsers/types.ts'
import type { AsyncEventEmitter } from './utils/AsyncEventEmitter.ts'
import type { FileManager } from './FileManager.ts'

declare global {
  namespace Kubb {
    interface App {}
  }
}

export type Component = any

export type AppEvents = {
  /**
   * Called in the beginning of the app lifecycle.
   */
  start: [{ app: App }]
  /**
   * Called in the end of the app lifecycle.
   */
  end: [{ app: App }]
  /**
   * Called when being rendered
   */
  render: [{ app: App }]
  /**
   * Called once before processing any files.
   */
  'process:start': [{ files: KubbFile.ResolvedFile[] }]

  /**
   * Called for each file when processing begins.
   */
  'file:start': [{ file: KubbFile.ResolvedFile; index: number; total: number }]

  /**
   * Called for each file when processing finishes.
   */
  'file:end': [{ file: KubbFile.ResolvedFile; index: number; total: number }]

  /**
   * Called periodically (or after each file) to indicate progress.
   * Useful for progress bars or logging.
   */
  'process:progress': [
    {
      processed: number
      total: number
      percentage: number
      source: string
      file: KubbFile.ResolvedFile
    },
  ]

  /**
   * Called once all files have been processed successfully.
   */
  'process:end': [{ files: KubbFile.ResolvedFile[] }]
}

export type AppContext<TOptions = unknown> = {
  options?: TOptions
  events: AsyncEventEmitter<AppEvents>
  fileManager: FileManager
  installedPlugins: Set<Plugin>
  installedParsers: Set<Parser>
}

export type Install<TOptions = any[] | object | undefined> = TOptions extends any[]
  ? (app: App, ...options: TOptions) => void
  : TOptions extends object
    ? (app: App, options?: TOptions) => void
    : (app: App) => void

export type Inject<TOptions = any[] | object | undefined, TAppExtension extends Record<string, any> = {}> = TOptions extends any[]
  ? (app: App, ...options: TOptions) => Partial<TAppExtension>
  : TOptions extends object
    ? (app: App, options?: TOptions) => Partial<TAppExtension>
    : (app: App) => Partial<TAppExtension>

export interface App<TOptions = unknown> extends Kubb.App {
  context: AppContext<TOptions>
  files: Array<KubbFile.ResolvedFile>
  use<TOptions extends any[] | object = any, TMeta extends object = object, TAppExtension extends Record<string, any> = {}>(
    pluginOrParser: Plugin<TOptions, TAppExtension> | Parser<TOptions, TMeta>,
    ...options: TOptions extends any[] ? NoInfer<TOptions> : [NoInfer<TOptions>]
  ): this & TAppExtension
  use<TOptions extends any[] | object = any, TMeta extends object = object, TAppExtension extends Record<string, any> = {}>(
    pluginOrParser: Plugin<TOptions, TAppExtension> | Parser<TOptions, TMeta>,
  ): this & TAppExtension
  addFile(...files: Array<KubbFile.File>): Promise<void>
}
