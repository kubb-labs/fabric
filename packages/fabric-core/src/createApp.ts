import { defineApp } from './defineApp.ts'

export const createApp = defineApp()
