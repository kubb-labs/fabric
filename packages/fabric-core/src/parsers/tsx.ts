import { typeScriptParser } from './typescript.ts'
import { createParser } from './createParser.ts'

export const tsxParser = createParser({
  name: 'tsx',
  install() {},
  async print(file, options = { extname: '.tsx' }) {
    return typeScriptParser.print(file, options)
  },
})
