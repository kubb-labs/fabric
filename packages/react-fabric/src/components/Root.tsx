import { Component, createContext } from 'react'

import type { KubbNode } from '../types.ts'

type ErrorBoundaryProps = {
  onError: (error: Error) => void
  children?: KubbNode
}

class ErrorBoundary extends Component<{
  onError: ErrorBoundaryProps['onError']
  children?: KubbNode
}> {
  state = { hasError: false }

  static displayName = 'KubbErrorBoundary'
  static getDerivedStateFromError(_error: Error) {
    return { hasError: true }
  }

  componentDidCatch(error: Error) {
    if (error) {
      this.props.onError(error)
    }
  }

  render() {
    if (this.state.hasError) {
      return null
    }
    return this.props.children
  }
}

export type RootContextProps = {
  /**
   * Exit (unmount) the whole Ink app.
   */
  readonly exit: (error?: Error) => void
}

export const RootContext = createContext<RootContextProps>({
  exit: () => {},
})

type RootProps = {
  /**
   * Exit (unmount) hook
   */
  readonly onExit: (error?: Error) => void
  /**
   * Error hook
   */
  readonly onError: (error: Error) => void
  readonly children?: KubbNode
}

export function Root({ onError, onExit, children }: RootProps) {
  try {
    return (
      <ErrorBoundary
        onError={(error) => {
          onError(error)
        }}
      >
        <RootContext.Provider value={{ exit: onExit }}>{children}</RootContext.Provider>
      </ErrorBoundary>
    )
  } catch (_e) {
    return null
  }
}

Root.Context = RootContext
Root.displayName = 'KubbRoot'
