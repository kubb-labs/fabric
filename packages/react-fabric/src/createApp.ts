import { defineApp } from '@kubb/fabric-core'
import { type ElementType, createElement } from 'react'
import { type ReactAppContext, ReactTemplate } from './ReactTemplate.tsx'

export const createApp = defineApp<ElementType, ReactAppContext>((app, context) => {
  const template = new ReactTemplate(context)

  return {
    render() {
      template.render(createElement(app))
    },
    async renderToString() {
      return template.renderToString(createElement(app))
    },
    waitUntilExit() {
      return template.waitUntilExit()
    },
  }
})
