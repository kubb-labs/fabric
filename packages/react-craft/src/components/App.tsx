import { createContext, useContext } from 'react'

import type { KubbFile } from '@kubb/craft-core'
import type { KubbNode } from '../types.ts'
import { RootContext } from './Root.tsx'

type AppContextProps = {
  /**
   * Exit (unmount)
   */
  readonly exit: (error?: Error) => void
  readonly mode: KubbFile.Mode
}

const AppContext = createContext<AppContextProps | undefined>(undefined)

type Props = {
  readonly mode: KubbFile.Mode
  readonly children?: KubbNode
}

export function App({  mode, children }: Props) {
  const { exit } = useContext(RootContext)

  return <AppContext.Provider value={{ exit,  mode }}>{children}</AppContext.Provider>
}

App.Context = AppContext
App.displayName = 'KubbApp'
