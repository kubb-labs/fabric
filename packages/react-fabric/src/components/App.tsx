import { createContext, useContext } from 'react'
import type { KubbNode } from '../types.ts'
import { RootContext } from './Root.tsx'

export type AppContextProps<TMeta = unknown> = {
  /**
   * Exit (unmount)
   */
  readonly exit: (error?: Error) => void
  readonly meta: TMeta
}

const AppContext = createContext<AppContextProps | undefined>(undefined)

type Props<TMeta = unknown> = {
  readonly children?: KubbNode
  readonly meta: TMeta
}

export function App<TMeta = unknown>({ meta, children }: Props<TMeta>) {
  const { exit } = useContext(RootContext)

  return <AppContext.Provider value={{ exit, meta }}>{children}</AppContext.Provider>
}

App.Context = AppContext
App.displayName = 'KubbApp'
