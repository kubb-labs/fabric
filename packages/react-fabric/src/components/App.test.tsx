import { createApp } from '../createApp.ts'
import { App } from './App.tsx'
import { Root } from './Root.tsx'

describe('<App/>', () => {
  test('render App with meta and children', async () => {
    const Component = () => {
      return (
        <Root onExit={() => {}} onError={() => {}}>
          <App meta={{ color: 'blue', version: 1 }}>
            AppChildren
            <App.Context.Consumer>{(ctx) => <>{`|meta:${JSON.stringify(ctx?.meta)}|exit:${typeof ctx?.exit}|`}</>}</App.Context.Consumer>
          </App>
        </Root>
      )
    }

    const app = createApp(Component)
    const output = await app.renderToString()

    expect(output).toMatchSnapshot()
  })
})
