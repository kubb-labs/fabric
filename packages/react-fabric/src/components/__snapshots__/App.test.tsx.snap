// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<App/> > render App with meta and children 1`] = `"AppChildren|meta:{"color":"blue","version":1}|exit:function|"`;
